File: backend/app/models/review.py
from sqlalchemy import Integer, Column, ForeignKey, SmallInteger, String, Text, Date, Enum, TIMESTAMP, func, \
    UniqueConstraint, Index, text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import relationship

from app.db.base import Base
from app.models.enums import ReviewStatus


class Review(Base):
    __tablename__ = "review"

    id = Column(Integer, primary_key=True, autoincrement=True)
    place_id = Column(Integer, ForeignKey('place.id', ondelete='CASCADE'), nullable=False)
    user_id = Column(Integer, ForeignKey('users.id', ondelete='CASCADE'), nullable=False)
    rating = Column(SmallInteger, nullable=False)
    title = Column(String(255))
    content = Column(Text)
    images = Column(JSONB)
    visit_date = Column(Date)
    status = Column(Enum(ReviewStatus, native_enum=False), server_default="APPROVED", nullable=False)
    created_at = Column(TIMESTAMP, nullable=False, server_default=func.now())
    updated_at = Column(TIMESTAMP, nullable=False, server_default=func.now())

    __table_args__ = (
        UniqueConstraint("place_id", "user_id"),
        Index('ix_review_user_id', 'user_id'),
        # calculate_place_rating(): AVG/COUNT theo place_id + status = APPROVED
        Index('ix_review_place_status', 'place_id', 'status'),
        # list_full() phía admin: filter status + ORDER BY created_at DESC
        Index('ix_review_status_created_at', 'status', text('created_at DESC')),
    )
    user = relationship('User', lazy=True)

File: mobile/lib/model/review.dart
class ReviewUser {
  final String name;
  final String? avatar;

  ReviewUser({required this.name, required this.avatar});

  factory ReviewUser.fromJson(Map<String, dynamic> json) => ReviewUser(
    name: json['name'] as String,
    avatar: json['avatar'] as String?,
  );
}

class Review {
  final int id;
  final int placeId;
  final int userId;
  final ReviewUser user;
  final int rating;
  final String? title;
  final String? content;
  final DateTime? visitDate;
  final List<String> images;
  final DateTime createdAt;

  Review({
    required this.id,
    required this.placeId,
    required this.userId,
    required this.user,
    required this.rating,
    required this.title,
    required this.content,
    required this.visitDate,
    required this.images,
    required this.createdAt,
  });

  
  factory Review.fromJson(Map<String, dynamic> json) => Review(
    id: json['id'] as int,
    placeId: json['place_id'] as int,
    userId: json['user_id'] as int,
    user: ReviewUser.fromJson(json['user'] as Map<String, dynamic>),
    rating: json['rating'] as int,
    title: json['title'] as String?,
    content: json['content'] as String?,
    visitDate: json['visit_date'] != null
        ? DateTime.parse(json['visit_date'] as String)
        : null,
    images: (json['images'] as List? ?? []).map((e) => e as String).toList(),
    createdAt: DateTime.parse(json['created_at'] as String),
  );
}
